sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolder.java

```diff
@@ -55,6 +61,11 @@ public BufferHolder(UnsafeRow row, int initialSize) {
    * Grows the buffer by at least neededSize and points the row to the buffer.
    */
   public void grow(int neededSize) {
+    if (neededSize > Integer.MAX_VALUE / 2 - totalSize()) {
+      throw new UnsupportedOperationException(
+        "Cannot grow BufferHolder by size " + neededSize + " because the size after growing " +
+          "exceeds size limitation " + Integer.MAX_VALUE / 2);
+    }
     final int length = totalSize() + neededSize;
     if (buffer.length < length) {
       // This will not happen frequently, because the buffer is re-used.
```

Comment:
@cloud-fan 

Currently the limit for `neededSize + totalSize` is `Integer.MAX_VALUE / 2`, I don't see there is a big difference to enlarge the limit to `Integer.MAX_VALUE`.

`Integer.MAX_VALUE / 2` is about 1 GB, it is quite rare for a single row to exceed this limit.